tap_marketo/sync.py

```diff
@@ -431,6 +434,170 @@ def sync_paginated(client, state, stream):
+    data = client.request("GET", endpoint)
+    record_count = 0
+
+    time_extracted = utils.now()
+    for row in data["result"]:
+        record = format_values(stream, row)
+        record_count += 1
+        singer.write_record(stream["tap_stream_id"], record, time_extracted=time_extracted)
+    return state, record_count
+
+def check_if_first_sync(state, tap_stream_id):
+    """
+    checks for initial sync flag stored in the bookmark
+    """
+    first_sync_state = singer.get_bookmark(state, tap_stream_id, INITAL_SYNC_KEY, None)
+    return True if first_sync_state in (None, False) else False
+
+def get_activate_version(state, tap_stream_id):
+    """
+        checks if the sync was interupted in-between an extraction
+        if interupted returns the previous activate version number used
+        else returns a new activate version
+    """
+
+    stream_version = singer.get_bookmark(state, tap_stream_id, ACTIVATE_VERSION_KEY, None)
+    if stream_version is None:
+        stream_version = int(time.time() * 1000)
+        singer.log_info("Using new activate version %s", stream_version)
+    else:
+        singer.log_info("Using interupted activate version %s", stream_version)
+    return stream_version
+
+def get_program_ids(client):
+    """
+    Gets a list of all programs to be used for the program_tags stream
+    """
+
+    params = {
+        "maxReturn": 200,
+        "offset": 0,
+    }
+    endpoint = "rest/asset/v1/programs.json"
+    rec_ids = []
+
+    while True:
+        data = client.request("GET", endpoint, endpoint_name="programs", params=params)
+        if "warnings" in data and NO_ASSET_MSG in data["warnings"]:
+            break
+        if "errors" in data and data["errors"] != []:
+            singer.log_info("error %s", data)
+            break
+        for rec in data["result"]:
+            rec_ids.append(rec["id"])
+        params["offset"] += params["maxReturn"]
+    return sorted(rec_ids)
+
+def sync_program_tags(client, state, stream):
+    singer.write_schema(stream["tap_stream_id"], stream["schema"], stream["key_properties"])
+    first_run = check_if_first_sync(state, stream["tap_stream_id"])
+
+    # check the interupted state of the program_id's list
+    last_synced_program = bookmarks.get_bookmark(state, stream["tap_stream_id"], "program_id")
+    stream_version = get_activate_version(state, stream["tap_stream_id"])
+
+    activate_version_message = singer.ActivateVersionMessage(
+            stream=stream['tap_stream_id'],version=stream_version)
+
+    if first_run:
+        # inital sync requires the activate version to be sent before sending records
+        singer.write_message(activate_version_message)
+        state = bookmarks.write_bookmark(state, stream["tap_stream_id"],INITAL_SYNC_KEY, True)
+        singer.write_state(state)
+
+    state = bookmarks.write_bookmark(state, stream["tap_stream_id"], ACTIVATE_VERSION_KEY, stream_version)
+    singer.write_state(state)
```

Comment:
We do not need to be writing state twice consecutively here

Reply:
Not necessary, since both statements are consecutive.
However this block will be executed only once (i.e the first sync), hence i made the state to be explicitly written in the same block.